include/symbol.hpp

```diff
@@ -16,18 +16,42 @@ struct SymbolEntry {
       : id{std::move(id)}, type{std::move(expr_type)} {}
 };
 
-class SymbolTable {
+struct DeclTypeEntry {
+  std::string id;
+  std::unique_ptr<Type> type;
+
+  DeclTypeEntry(std::string id, std::unique_ptr<Type> type)
+      : id{std::move(id)}, type{std::move(type)} {}
+};
+
+template <typename EntryType>
+class TableTemplate {
  public:
   /// @brief Adds the `entry` to the table if the `id` of the `entry` isn't
   /// already in the table.
   /// @returns The added entry if the `id` of the `entry` isn't
   /// already in the table; otherwise, the original entry.
-  std::shared_ptr<SymbolEntry> Add(std::unique_ptr<SymbolEntry> entry);
-
-  std::shared_ptr<SymbolEntry> Probe(const std::string& id) const;
+  std::shared_ptr<EntryType> Add(std::unique_ptr<EntryType> entry) {
+    const std::string& id = entry->id;  // to reference id after moved
+    if (!Probe(id)) {
+      entries_.insert({id, std::shared_ptr<EntryType>{std::move(entry)}});
+    }
+    return entries_.at(id);
+  }
+  std::shared_ptr<EntryType> Probe(const std::string& id) const {
+    if (entries_.count(id)) {
+      return entries_.at(id);
+    }
+    return nullptr;
+  }
 
  private:
-  std::map<std::string, std::shared_ptr<SymbolEntry>> entries_{};
+  std::map<std::string, std::shared_ptr<EntryType>> entries_{};
 };
 
+/// @brief Stores identifiers.
+using SymbolTable = TableTemplate<SymbolEntry>;
```

Comment:
I'm a bit confused about the difference between _identifier_ and _symbol_. The `DeclTypeEntry` also holds a `std::string id`, so I believe `id` is a more general concept. How about we have _symbol_ refer to variable ids and function ids? 

As a result, `DeclType` can naturally refer to the ids that declare a type. This way, the ambiguous comment can be removed.

```suggestion
using SymbolTable = TableTemplate<SymbolEntry>;
```